### ISSUE TITLE
boost python namespace conflict

### ISSUE THREAD

**Issue by picca** [cctbx/cctbx_project#458]
Hello, I am working on a Debian package for dials and cctbx.

but when I tryed to install the package, I got this error

```
~/Debian/dials/dials$ LANG=C sudo debi
(Reading database ... 474720 files and directories currently installed.)
Preparing to unpack dials-doc_2.2.1+dfsg2-1_all.deb ...
Unpacking dials-doc (2.2.1+dfsg2-1) over (2.2.1+dfsg2-1) ...
Preparing to unpack dials_2.2.1+dfsg2-1_amd64.deb ...
Unpacking dials (2.2.1+dfsg2-1) over (2.2.1+dfsg2-1) ...
Preparing to unpack python3-dials_2.2.1+dfsg2-1_amd64.deb ...
Unpacking python3-dials (2.2.1+dfsg2-1) ...
dpkg: error processing archive python3-dials_2.2.1+dfsg2-1_amd64.deb (--install):
 trying to overwrite '/usr/lib/python3/dist-packages/boost/__init__.py', which is also in package libboost-mpi-python1.67.0 1.67.0-17+b1
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Setting up dials-doc (2.2.1+dfsg2-1) ...
dpkg: dependency problems prevent configuration of dials:
 dials depends on python3-dials; however:
  Package python3-dials is not installed.

dpkg: error processing package dials (--install):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 python3-dials_2.2.1+dfsg2-1_amd64.deb
 dials
debi: error: debpkg -i dials-doc_2.2.1+dfsg2-1_all.deb dials_2.2.1+dfsg2-1_amd64.deb python3-dials_2.2.1+dfsg2-1_amd64.deb subprocess returned exit status 1
```

Indeed boost-mpi-python already provide a boost package. So there is a conflict between cctbx and boost-mpi-python.

It seems to me that you should not preempt the boost namespace and rename your boost package with another name or move it somewhere else below cctbx maybe.

thanks for considering.

**Comment by phyy-nx** [comment]
Thanks for pointing this out.  Discussed this with @bkpoon this afternoon.  To summarize, the following current use case is not ideal as we shouldn't be claiming the `boost` namespace:

```
$ conda install -c cctbx-dev cctbx
$ python
>>> import boost 
```
(`import boost` imports our module named `boost` which includes convenience wrappers for loading our boost modules)

So we propose reverting part of 032d4e1, namely moving cctbx_project/boost back to cctbx_project/boost_adaptbx.  However, this will require the following code changes in **EVERY** package that uses cctbx, including (at least) phenix, phaser, dials, dxtbx, elbow:

1. `import boost.python` --> `import boost_adaptbx.boost.python`: in cctbx_project that's 239 places
2. `boost.python.import_ext` --> `boost_adaptbx.boost.python.import_ext`: in cctbx_project that's 228 places
3. `@boost.python.inject_into` --> `@boost_adaptbx.boost.python.inject_into`: 114 places

Plus additional changes not yet identified where the `boost` module is imported or used.

**Comment by Anthchirp** [comment]
Suggest moving the files from `boost/` into `boost_adaptbx/` instead of `boost_adaptbx/boost/`.

Since it looks that this will involve some coordination effort I would also suggest we leave redirection stubs in place that raise first `DeprecationWarning`s, then `UserWarning`s to help in the transition period before removing the old names entirely.

**Comment by graeme-winter** [comment]
+1 to idea of moving files rather than the additional layer of `boost`

**Comment by picca** [comment]
Hello, 

what about moving all this boost  + adapt_boost in the cctbx namespace under

cctbx.boost

this looks simple and  the meaning is clear, this is something related to boost from the cctbx community.
It is easy to type and remember.

my two cents :)

**Comment by Anthchirp** [comment]
This would be a good idea if everything in cctbx_project was cctbx. But unfortunately that isn't the case 🙂

**Comment by picca** [comment]
that would be my next question :).

why don't you move everythings under your namespace cctbx.
It would be lot clearer for eveytone that this is comming from the cctbx community.
And this way you minimize this namespace collision.

**Comment by phyy-nx** [comment]
From @Anthchirp:

> Suggest moving the files from boost/ into boost_adaptbx/ instead of boost_adaptbx/boost/.
> Since it looks that this will involve some coordination effort I would also suggest we leave redirection stubs in place that raise first DeprecationWarnings, then UserWarnings to help in the transition period before removing the old names entirely.

Agree with deprecations.  @bkpoon is considering renaming the `boost` directory to `python` and moving the stuff currently in `boost/python.py` into `boost/__init__.py`.  So you would `import boost_adaptbx.python` to call `import_ext`.

@picca moving everything e.g. iotbx, xfel, rstbx, mmtbx, and so forth under the cctbx would be a ton of work and would break a lot of things.  I think that's outside the scope of what we can do realistically for now.

**Comment by Anthchirp** [comment]
So you mean
`boost/__init__.py` -> `boost_adaptbx/python/__init__.py`
`boost/python.py` -> `boost_adaptbx/python/__init__.py` (concatenated)

`boost/optional.py` -> `boost_adaptbx/python/optional.py`
`boost/rational.py` -> `boost_adaptbx/python/rational.py`
`boost/std_pair.py` -> `boost_adaptbx/python/std_pair.py`
`boost/tuple.py` -> `boost_adaptbx/python/tuple.py`
?

(Happy to create a PR with those changes if you like)

**Comment by phyy-nx** [comment]
Yah, that's what I mean.

Let's get some more input before the PR.  @luc-j-bourhis, @pcxod can you opine on this?  Thanks.

**Comment by bkpoon** [comment]
@picca This was something that was on my list of things to discuss as I was building the `conda` package for `cctbx` and would like this resolved before building the package on `conda-forge`.

@Anthchirp @phyy-nx I've taken a closer look at the items in the `boost` directory and it looks like we were trying to have something that maps roughly to the layout of Boost when exposing things to Python. So `boost/optional.py` maps to Boost.Optional, `boost/python.py` maps to Boost.Python, `boost/rational.py` maps to `Boost.Rational`, and `boost/tuple.py` maps to `Boost.Tuple`. It seems like my idea of moving things to `boost_adaptbx/python` doesn't quite fit and keeping the files in `boost_adaptbx/boost` should be fine. Also, `boost_adaptbx/graph` was added later. Do @luc-j-bourhis or @pcxod have any ideas or preferences since this change should affect you as well?

**Comment by luc-j-bourhis** [comment]
@bkpoon I think that simply moving `boost` to `boost_adaptbx` is the best solution indeed. Since `smtbx` is doing the same as all other modules, I don't foresee it being any special case at all.

**Comment by Anthchirp** [comment]
I've prepared a pull request in #470. This would be stage 1 - moving code and marking it as deprecated in the previous location.

I hold no opinion how long the old `boost.*` names should remain available.

To then close this issue all that needs doing is stage 2: delete the `boost/` directory.

**Comment by bkpoon** [comment]
Fixed with #470 and #516.
